cmd/soroban-cli/src/rpc/transaction.rs

```diff
@@ -78,12 +79,160 @@ pub fn assemble(
+                ..
+            } = auth else {
+                // Doesn't need special signing
+                return Ok(auth);
+            };
+            let SorobanAddressCredentials { ref address, .. } = credentials;
+
+            // See if we have a signer for this authorizationEntry
+            // If not, then we Error
+            let needle = match address {
+                ScAddress::Account(AccountId(PublicKey::PublicKeyTypeEd25519(Uint256(ref a)))) => a,
+                ScAddress::Contract(Hash(c)) => {
+                    // This address is for a contract. This means we're using a custom
+                    // smart-contract account. Currently the CLI doesn't support that yet.
+                    return Err(Error::MissingSignerForAddress {
+                        address: stellar_strkey::Strkey::Contract(stellar_strkey::Contract(*c))
+                            .to_string(),
+                    });
+                }
+            };
+            let signer = if let Some(s) = signers.iter().find(|s| needle == s.public.as_bytes()) {
+                s
+            } else if needle == source_address {
+                // This is the source address, so we can sign it
+                source_key
+            } else {
+                // We don't have a signer for this address
+                return Err(Error::MissingSignerForAddress {
+                    address: stellar_strkey::Strkey::PublicKeyEd25519(
+                        stellar_strkey::ed25519::PublicKey(*needle),
+                    )
+                    .to_string(),
+                });
+            };
+
+            sign_soroban_authorization_entry(
+                raw_auth,
+                signer,
+                signature_expiration_ledger,
+                &network_id,
+            )
+        })
+        .collect::<Result<Vec<_>, Error>>()?;
+
+    body.auth = signed_auths.clone().try_into()?;
+    tx.operations = vec![op].try_into()?;
+    Ok((tx, signed_auths))
+}
+
+pub fn sign_soroban_authorization_entry(
+    raw: &SorobanAuthorizationEntry,
+    signer: &ed25519_dalek::Keypair,
+    signature_expiration_ledger: u32,
+    network_id: &Hash,
+) -> Result<SorobanAuthorizationEntry, Error> {
+    let mut auth = raw.clone();
+    let SorobanAuthorizationEntry {
+        credentials: SorobanCredentials::Address(ref mut credentials),
+        ..
+    } = auth else {
+        // Doesn't need special signing
+        return Ok(auth);
+    };
+    let SorobanAddressCredentials { nonce, .. } = credentials;
+
+    let preimage = HashIdPreimage::SorobanAuthorization(HashIdPreimageSorobanAuthorization {
+        network_id: network_id.clone(),
+        invocation: auth.root_invocation.clone(),
+        nonce: *nonce,
+        signature_expiration_ledger,
+    })
+    .to_xdr()?;
+
+    let payload = Sha256::digest(preimage);
+    let signature = signer.sign(&payload);
+
+    let map = ScMap::sorted_from(vec![
```

Comment:
why sorted?

Reply:
Only sorted `ScMap`s are considered valid by the Soroban host.

Reply:
oh what?? good to know :memo: 

but the keys are sorted here anyway so it'd be a waste

Reply:
In general, `ScMap`s should be created with `sorted_from`. I'm not even sure if we provide a different constructor. We shouldn't try to optimize something as marginal as sorting a couple of keys, while risking to break compatibility with the host.

Also, I would be much happier if we didn't build `ScVal`s manually at all and used SDK here. That would come with some additional deps and somewhat higher costs, but the benefit would be guaranteed compatibility with host and much more natural structure building. Maybe we should think how we can expose SDK+host for the Rust clients without bringing all the deps...